File: Day13/Java/Main.java
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Main {

	public static void main(String[] args) throws IOException {
		File input = new File("input").getAbsoluteFile();
		if (!input.exists()) {
			System.out.println(input.getPath() + " does not exist!");
			System.out.println("trying " + input.getPath() + ".txt instead.");
			input = new File(input.getPath() + ".txt");
			if (!input.exists()) {
				System.err.println("neither " + input.getPath().substring(0, input.getPath().length() - 4) + ",");
				System.err.println("nor " + input.getPath() + " exists!");
				return;
			}
		}

		int departing = 0;
		List<Integer> busIds = new ArrayList<>();
		Map<Integer, Integer> busses = new HashMap<>();
		int fastest = 0;
		for (String line : Files.readAllLines(input.toPath())) {
			if (departing == 0) {
				departing = Integer.parseInt(line);
			} else {
				for (String bus : line.split(",")) {
					if (!bus.equals("x")) {
						int b = Integer.parseInt(bus);
						busses.put(b, b - departing % b);
						if (fastest == 0 || busses.get(b) < busses.get(fastest)) {
							fastest = b;
						}
						busIds.add(b);
					} else
						busIds.add(0);
				}
			}
		}

		System.out.println("Part one result is " + fastest * busses.get(fastest) + ".");

		long time = 0;
		long step = 1;
		int current = -1;
		while (true) {
			int bus = 0;
			for (int i = current + 1; i < busIds.size(); i++) {
				bus = busIds.get(i);
				if (bus == 0) {
					continue;
				} else if ((time + i) % bus != 0) {
					bus = 0;
					break;
				} else {
					current = i;
					step *= bus;
				}
			}

			if (bus == busIds.get(busIds.size() - 1))
				break;
			else
				time += step;
		}

		System.out.println("Part two timestamp is " + time + ".");
	}
}

File: Day13/Python/Main.py
import os

def main():
    input = os.path.dirname(os.getcwd())
    input = os.path.join(input, "input")
    if not os.path.exists(input):
        print(input + " does not exist!")
        input = os.path.join(os.path.dirname(input), "input.txt")
        print(f"trying {input} instead.")
        if not os.path.exists(input):
            print(input + " does not exist!")
            return;

    inputFile = open(input, 'r')
    departing = 0
    busIds = []
    busses = {}
    fastest = 0;
    for line in inputFile.readlines():
        if departing == 0:
            departing = int(line[:-1])
        else:
            for bus in line.split(','):
                if bus != "x":
                    b = int(bus)
                    busses[b] = b - departing % b
                    if fastest == 0 or busses[b] < busses[fastest]:
                        fastest = b
                    busIds.append(b)

                else:
                    busIds.append(0)

    print(f"Part one result is {fastest * busses[fastest]}.")

    time = 0
    step = 1
    current = -1
    while(True):
        bus = 0
        for i in range(current + 1, len(busIds)):
            bus = busIds[i]
            if bus == 0:
                continue
            elif (time + i) % bus != 0:
                bus = 0
                break;
            else:
                current = i
                step *= bus

        if bus == busIds[-1]:
            break
        else:
            time += step

    print(f"Part two timestamp is {time}.")

if __name__ == '__main__':
    main()
